darts/timeseries.py

```diff
@@ -1928,6 +1915,139 @@ def _combine_arrays(self,
+        for more info on supported frequencies) from the provided timestamps.
+
+        Parameters
+        ----------
+        xa
+            The xarray DataArray
+        freq
+            Optionally, a string representing the frequency of the Pandas DateTimeIndex to fill in the missing dates.
+
+        Raises
+        -------
+        ValueError
+            If `xa`'s DateTimeIndex contains less than 3 elements;
+            if no unique frequency can be inferred from `xa`'s DateTimeIndex;
+            if the resampled DateTimeIndex does not contain all dates from `xa` (see
+                :meth:`_resample_xarray() <TimeSeries._resample_xarray>`)
+
+        Returns
+        -------
+        xarray DataArray
+            xarray DataArray with filled missing dates from `xa`
+        """
+
+        if freq is not None:
+            return TimeSeries._resample_xarray(xa, freq, sort=True)
+
+        raise_if(len(xa) <= 2, f"Input time series must be of (length>=3) when fill_missing_dates=True and freq=None.",
+                 logger)
+
+        sorted_xa = xa.sortby(xa.dims[0])
+        time_index = sorted_xa.get_index(xa.dims[0])
+
+        offset_alias_info = "For more information about frequency aliases, read " \
+                            "https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#offset-aliases"
+
+        n_tests = 10
+        # sizes 3: high probability to find an inferred frequency, 10: recognize larger patterns (i.e. business hours)
+        samples_sizes = [3, 10]
+        observed_freqs = set()
+        # if we cannot infer a frequency within the first 20 dates, the input data is too chaotic anyway.
+        # within the first 20 dates, all possible frequencies (offset aliases) should be identified
+        for samples_size in samples_sizes:
+            for i in range(n_tests):
+                inferred_freq = time_index[i:i + samples_size].inferred_freq
+                if inferred_freq not in observed_freqs and inferred_freq is not None:
+                    observed_freqs.add(inferred_freq)
+
+        raise_if_not(
+            len(observed_freqs) == 1,
+            f"Could not observe an inferred frequency from the first n_first<={samples_sizes[-1] + n_tests} "
+            f"time stamps. An explicit frequency must be evident over a span of at least 3 consecutive time "
+            f"stamps within the first n_first dates. {offset_alias_info}" if not len(observed_freqs) else
+            f"Could not find a unique inferred frequency (not constant). Observed frequencies: {observed_freqs}. "
+            f"If any of those is the actual frequency, try passing it with fill_missing_dates=True "
+            f"and freq=your_frequency. {offset_alias_info}",
+            logger)
+
+        freq = observed_freqs.pop()
+
+        return TimeSeries._resample_xarray(sorted_xa, freq, sort=False)
+
+    @staticmethod
+    def _resample_xarray(xa: xr.DataArray,
```

Comment:
Yes and no. 

- Why not: `xa.resample(...).asfreq()` can upsample and downsample, whereas `_resample_xarray()` tries to re/create the array with a filled DatetimeIndex that has the same/inferred frequency as the input array and which **must** contain all indices and data from the input array plus the missing dates filled with NaNs. 

From what I understood, when creating a new TimeSeries, we do not want to resample to a different frequency as the input array, is that correct? And if someone wants to resample, we have `TimeSeries.resample()` for that.

- Why yes: `_resample_xarray()` is much faster than`xa.resample(...).asfreq()` as you can see below (blue is actually that low).
- For resampling to the unique inferred frequency, the output of both methods is the same. 

![image](https://user-images.githubusercontent.com/84317528/134517475-aef09b38-71d5-43e1-a2f4-dff11cac9fca.png)

Reply:
OK so if I understood well it is doing the same thing when resampling with the unique inferred frequency (which is how we are using the method), but also this one is much faster. That's great!!

Just curiosity question : do you know what's the performance hit of creating TimeSeries with `fill_missing_dates=True` now, with this performance improvement (I expect it'll still be slow because of the frequency inference part, but I'm wondering how slow).

Reply:
I think a different name for the new resample method would be good to avoid confusion.
maybe `rebuild_xarray_from_frequency` or `restore_xarray_from_frequency` ?

Reply:
Good idea, I vote for `_restore_xarray_from_frequency()`

Reply:
About performance hit, I can compare it to the `energy_dataset.csv` with shape (~40k, 30). Before it needed 8-9 seconds and now it needs ~2.3 seconds. Not very sophisticated comparison, I know :)

Reply:
could do some more tests if you want?

Reply:
I think it's good enough for now, we can perhaps do more advanced benchmarks another time. But now we can already close this PR, it's good progress :)